src/region_merging.jl

```diff
@@ -0,0 +1,111 @@
+makebt(i) = ()
+@inline makebt(i, ind, rest...) = (i & 1 != 0, makebt(i>>1, rest...)...)
+
+makert(ind::Int, start_ind::NTuple, mid_ind::NTuple, end_ind::NTuple) = ()
+@inline makert(ind::Int, start_ind::NTuple, mid_ind::NTuple, end_ind::NTuple, b::Bool, rest...) =
+    (b ? (mid_ind[ind]+1:end_ind[ind]) : (start_ind[ind]:mid_ind[ind]), makert(ind+1, start_ind, mid_ind, end_ind, rest...)...)
+
+function region_tree!(rtree::Cell, img::AbstractArray{T,N}, homogeneous::Function) where T<:Union{Colorant, Real} where N
+
+    if *(length.(indices(img))...) == 0
+        return rtree
+    end
+
+    if homogeneous(img)
+        m = mean(img)
+        c = length(linearindices(img))
+        rtree.data = ((m,c))
+        return rtree
+    end
+
+    split!(rtree)
+    start_ind = first(CartesianRange(indices(img))).I
+    end_ind = last(CartesianRange(indices(img))).I
+    mid_ind = (start_ind.+end_ind).÷2
+
+    for i in 0:2^N-1
+        bt = makebt(i, start_ind...)
+        rt = makert(1, start_ind, mid_ind, end_ind, bt...)
+        region_tree!(rtree[(Int.(bt) .+ 1)...], view(img, rt...), homogeneous)
+    end
+    rtree
+end
+
+"""
+    t = region_tree(img, homogeneous)
+
+Creates a region tree from `img` by splitting it recursively until
+all the regions are homogeneous.
+
+    b = homogeneous(img)
+
+Returns true if `img` is homogeneous.
+
+# Examples
+
+```jldoctest
+# img is an array with elements of type `Float64`
+julia> homogeneous(img::AbstractArray{T,N}) where {T<:Colorant,N} = -(extrema(img)...)*-1 < Gray{Float64}(0.2);
```

Comment:
I think this would be clearer as
```julia
function homogeneous(img)
    min, max = extrema(img)
    max - min < 0.2
end
```
(I assume that works, right? If not, we should fix whatever is broken about it.)

Likewise below.